File: code/richediting/src/main/java/org/fbme/ide/richediting/adapters/ecc/cell/ActionBlock.java
package org.fbme.ide.richediting.adapters.ecc.cell;

import org.fbme.lib.iec61499.ecc.StateAction;

public class ActionBlock {
    private final AlgorithmCell myAlgorithm;
    private final OutputCell myOutput;
    private final StateAction myAction;

    public ActionBlock(AlgorithmCell algorithm, OutputCell output, StateAction action) {
        this.myAlgorithm = algorithm;
        this.myOutput = output;
        this.myAction = action;
    }

    public ActionCell getAlgorithm() {
        return myAlgorithm;
    }

    public ActionCell getOutput() {
        return myOutput;
    }

    public void relayout() {
        myAlgorithm.relayoutAll();
        myOutput.relayout();
    }

    public int newWidth(int oldWidth) {
        int maxTmp = Math.max(myAlgorithm.getAllWidth(), myOutput.getWidth());
        return Math.max(oldWidth, maxTmp);
    }

    public void setWidth(int width) {
        myAlgorithm.setAllWidth(width);
        myOutput.setWidth(width);
    }

    public int getHeight(int padding) {
        return myAlgorithm.getAllHeight() + myOutput.getHeight() + 2 * padding;
    }

    public int moveTo(int dx, int dy, int padding, int currentHeight) {
        myAlgorithm.moveToAll(dx, dy, currentHeight);
        currentHeight += (myAlgorithm.getAllHeight() + padding);

        myOutput.moveTo(dx, dy + currentHeight);
        currentHeight += (myOutput.getHeight() + padding);
        return currentHeight;
    }

    public StateAction getAction() {
        return myAction;
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/adapters/ecc/cell/ActionBlock.kt
package org.fbme.ide.richediting.adapters.ecc.cell

import kotlin.math.max

class ActionBlock(val algorithm: AlgorithmCell, val output: OutputCell) {
    fun relayout() {
        algorithm.relayoutAll()
        output.relayout()
    }

    fun newWidth(oldWidth: Int): Int {
        val maxTmp = max(algorithm.totalWidth, output.width)
        return max(oldWidth, maxTmp)
    }

    fun setWidth(width: Int) {
        algorithm.totalWidth = width
        output.width = width
    }

    fun getHeight(padding: Int): Int {
        return algorithm.totalHeight + output.height + 2 * padding
    }

    fun moveTo(dx: Int, dy: Int, padding: Int, currentHeight: Int): Int {
        var h = currentHeight
        algorithm.moveToAll(dx, dy, h)
        h += algorithm.totalHeight + padding
        output.moveTo(dx, dy + h)
        h += output.height + padding
        return h
    }
}
